packages/core-p2p/src/network-monitor.ts

```diff
@@ -301,57 +300,113 @@ export class NetworkMonitor implements P2P.INetworkMonitor {
+
+                for (peer of peersToTry) {
+                    blocks = await this.communicator.downloadBlocks(peer, height);
+
+                    if (blocks.length === chunkSize || (isLastChunk && blocks.length > 0)) {
+                        this.logger.debug(
+                            `Downloaded blocks ${blocksRange} (${blocks.length}) ` +
+                            `from ${peer.ip}:${peer.port}`
+                        );
+                        downloadResults[i] = blocks;
+                        return;
+                    }
 
-            if (!networkHeight || networkHeight <= fromBlockHeight) {
-                // networkHeight is what we believe network height is, so even if it is <= our height, we download blocks
-                return this.communicator.downloadBlocks(sample(peersFiltered), fromBlockHeight);
-            }
+                    if (someJobFailed) {
+                        throw new Error(`Prematurely canceled due to another job failure.`);
+                    }
+                }
 
-            const chunkSize: number = 400;
-            const chunksMissingToSync: number = Math.ceil((networkHeight - fromBlockHeight) / chunkSize);
-            const chunksToDownload: number = Math.min(chunksMissingToSync, peersFiltered.length, maxParallelDownloads);
+                someJobFailed = true;
 
-            return (await Promise.all(
-                shuffle(peersFiltered)
-                    .slice(0, chunksToDownload)
-                    .map(async (peer: P2P.IPeer, index) => {
-                        const height: number = fromBlockHeight + chunkSize * index;
-                        const peersToTry: P2P.IPeer[] = [peer, sample(peersFiltered), sample(peersFiltered)]; // 2 "fallback" peers to download from if 1st one failed
+                throw new Error(
+                    `Could not download blocks ${blocksRange} from any of ${peersToTry.length} ` +
+                    `peer(s). Last attempt returned ${blocks.length} block(s) from peer ${peer.ip}.`
+                );
+            });
 
-                        let blocks: Interfaces.IBlockData[];
-                        for (const peerToDownloadFrom of peersToTry) {
-                            blocks = await this.communicator.downloadBlocks(peerToDownloadFrom, height);
+            if (chunksHumanReadable.length > 0) {
+                chunksHumanReadable += ", ";
+            }
+            chunksHumanReadable += blocksRange;
+        }
 
-                            if (blocks.length > 0) {
-                                return blocks;
-                            }
-                        }
+        this.logger.debug(`Downloading blocks in chunks: ${chunksHumanReadable}`);
 
-                        return blocks;
-                    }),
-            )).reduce((acc, curr) => [...acc, ...(curr || [])], []);
+        let firstFailureMessage: string;
+
+        try {
+            await Promise.all(downloadJobs.map(f => f()));
```

Comment:
The `.map(f => f())`  can be removed, since `Promise.all` does that internally.

Reply:
That is necessary to convert the async functions to promises by calling them.